contracts/core/Registry.sol

```diff
@@ -1,47 +1,242 @@
+    /// ==== External/Public Functions =====
+    /// ====================================
+
+    /// @notice Retrieve identity by identityId
+    /// @param identityId The identityId of the identity
+    function getIdentityById(bytes32 identityId) public view returns (Identity memory) {
+        return identitiesById[identityId];
+    }
+
+    /// @notice Retrieve identity by anchor
+    /// @param anchor The anchor of the identity
+    function getIdentityByAnchor(address anchor) public view returns (Identity memory) {
+        bytes32 identityId = anchorToIdentityId[anchor];
+        return identitiesById[identityId];
+    }
+
+    /// @notice Creates a new identity
+    /// @dev This will also set the attestation address generated from msg.sender and name
+    /// @param _nonce Nonce used to generate identityId
+    /// @param _name The name of the identity
+    /// @param _metadata The metadata of the identity
+    /// @param _members The members of the identity
+    /// @param _owner The owner of the identity
+    function createIdentity(
+        uint256 _nonce,
+        string memory _name,
+        Metadata memory _metadata,
+        address _owner,
+        address[] memory _members
+    ) external returns (bytes32) {
+        bytes32 identityId = _generateIdentityId(_nonce);
+
+        if (identitiesById[identityId].nonce != 0) {
+            revert NONCE_NOT_AVAILABLE();
+        }
+
+        Identity memory identity = Identity({
+            nonce: _nonce,
+            owner: _owner,
+            name: _name,
+            metadata: _metadata,
+            anchor: _generateAnchor(identityId, _name)
+        });
+
+        identitiesById[identityId] = identity;
+        anchorToIdentityId[identity.anchor] = identityId;
+
+        // assign roles
+        uint256 memberLength = _members.length;
+        for (uint256 i = 0; i < memberLength;) {
+            _grantRole(identityId, _members[i]);
+            unchecked {
+                i++;
+            }
+        }
+
+        emit IdentityCreated(
+            identityId, identity.nonce, identity.name, identity.metadata, identity.owner, identity.anchor
+        );
+
+        return identityId;
+    }
+
+    /// @notice Updates the name of the identity and generates new anchor
+    /// @param _identityId The identityId of the identity
+    /// @param _name The new name of the identity
+    /// @dev Only owner can update the name.
+    function updateIdentityName(bytes32 _identityId, string memory _name)
+        external
+        isIdentityOwner(_identityId)
+        returns (address)
+    {
+        address anchor = _generateAnchor(_identityId, _name);
+
+        Identity storage identity = identitiesById[_identityId];
+        identity.name = _name;
+
+        // clear old anchor
+        anchorToIdentityId[identity.anchor] = bytes32(0);
+
+        // set new anchor
+        identity.anchor = anchor;
+        anchorToIdentityId[identity.anchor] = _identityId;
```

Comment:
Since `identity` is in storage, you can save gas by using `anchor` here instead of `identity.anchor`.